WordPressEditor/src/main/java/org/wordpress/android/editor/HtmlStyleTextWatcher.java

```diff
@@ -0,0 +1,196 @@
+    private Operation mLastOperation;
+
+    @Override
+    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+        int lastCharacterLocation = start + count - 1;
+        if (s.length() > lastCharacterLocation && lastCharacterLocation >= 0) {
+            if (after < count) {
+                if (after > 0) {
+                    // Text was deleted and replaced by some other text
+                    mLastOperation = Operation.REPLACE;
+                } else {
+                    // Text was deleted only
+                    mLastOperation = Operation.DELETE;
+                }
+
+                mOffset = start;
+                mModifiedText = s.subSequence(start + after, start + count);
+            }
+        }
+    }
+
+    @Override
+    public void onTextChanged(CharSequence s, int start, int before, int count) {
+        int lastCharacterLocation = start + count - 1;
+        if (s.length() > lastCharacterLocation) {
+            if (count > 0) {
+                if (before > 0) {
+                    // Text was added, replacing some existing text
+                    mLastOperation = Operation.REPLACE;
+                    mModifiedText = s.subSequence(start, start + count);
+                } else {
+                    // Text was added only
+                    mLastOperation = Operation.INSERT;
+                    mOffset = start;
+                    mModifiedText = s.subSequence(start + before, start + count);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void afterTextChanged(Editable s) {
+        if (mModifiedText == null) {
+            return;
+        }
+
+        SpanRange spanRange;
+
+        // If the modified text included a tag or entity symbol ("<", ">", "&" or ";"), find its match and restyle
+        if (mModifiedText.toString().contains("<")) {
+            spanRange = restyleForChangedOpeningSymbol(s, "<");
+        } else if (mModifiedText.toString().contains(">")) {
+            spanRange = restyleForChangedClosingSymbol(s, ">");
+        } else if (mModifiedText.toString().contains("&")) {
+            spanRange = restyleForChangedOpeningSymbol(s, "&");
+        } else if (mModifiedText.toString().contains(";")) {
+            spanRange = restyleForChangedClosingSymbol(s, ";");
+        } else {
+            // If the modified text didn't include any tag or entity symbols, restyle if the modified text is inside
+            // a tag or entity
+            spanRange = restyleNormalTextIfWithinSymbols(s, "<", ">");
+            if (spanRange == null) {
+                spanRange = restyleNormalTextIfWithinSymbols(s, "&", ";");
+            }
+        }
+
+        if (spanRange != null) {
+            updateSpans(s, spanRange);
+        }
+
+        mModifiedText = null;
+        mLastOperation = Operation.NONE;
+    }
+
+    protected SpanRange restyleForChangedOpeningSymbol(Editable content, String openingSymbol) {
```

Comment:
I won't leave any other comments about javadoc'ing methods so I don't clutter up this PR. :-)